src/ore/src/metrics/delete_on_drop.rs

```diff
@@ -0,0 +1,403 @@
+    _phantom: &'a PhantomData<()>,
+}
+
+impl<'a, P, L> DeleteOnDropCounter<'a, P, L>
+where
+    P: Atomic,
+    L: PromLabelsExt<'a>,
+{
+    fn from_metric_vector(vec: &GenericCounterVec<P>, labels: L) -> Self {
+        let inner = labels.get_from_metric_vec(vec);
+        Self {
+            inner,
+            labels,
+            vec: vec.clone(),
+            _phantom: &PhantomData,
+        }
+    }
+}
+
+impl<'a, P, L> Deref for DeleteOnDropCounter<'a, P, L>
+where
+    P: Atomic,
+    L: PromLabelsExt<'a>,
+{
+    type Target = GenericCounter<P>;
+    fn deref(&self) -> &GenericCounter<P> {
+        &self.inner
+    }
+}
+
+impl<'a, P, L> Drop for DeleteOnDropCounter<'a, P, L>
+where
+    P: Atomic,
+    L: PromLabelsExt<'a>,
+{
+    fn drop(&mut self) {
+        if self.labels.remove_from_metric_vec(&self.vec).is_err() {
+            // ignore.
+        }
+    }
+}
+
+/// Extension trait for all gauge metrics vectors.
+///
+/// It adds a method to create a concrete metric from the vector that gets removed from the vector
+/// when the concrete metric is dropped.
+pub trait CounterVecExt {
+    /// The type of value that the counter should count.
+    type CounterType: Atomic;
+
+    /// Returns a counter that deletes its labels from this metrics vector when dropped.
+    /// See [`DeleteOnDropCounter`] for a detailed description.
+    fn get_delete_on_drop_counter<'a, L: PromLabelsExt<'a>>(
+        &self,
+        labels: L,
+    ) -> DeleteOnDropCounter<'a, Self::CounterType, L>;
+}
+
+impl<P: Atomic> CounterVecExt for GenericCounterVec<P> {
+    type CounterType = P;
+
+    fn get_delete_on_drop_counter<'a, L: PromLabelsExt<'a>>(
+        &self,
+        labels: L,
+    ) -> DeleteOnDropCounter<'a, Self::CounterType, L> {
+        DeleteOnDropCounter::from_metric_vector(self, labels)
+    }
+}
+
+/// A [`GenericGauge`] wrapper that deletes its labels from the vec when it is dropped
+#[derive(Debug)]
+pub struct DeleteOnDropGauge<'a, P, L>
+where
+    P: Atomic,
+    L: PromLabelsExt<'a>,
+{
+    inner: GenericGauge<P>,
+    labels: L,
+    vec: GenericGaugeVec<P>,
+    _phantom: &'a PhantomData<()>,
+}
+
+impl<'a, P, L> DeleteOnDropGauge<'a, P, L>
+where
+    P: Atomic,
+    L: PromLabelsExt<'a>,
+{
+    fn from_metric_vector(vec: &GenericGaugeVec<P>, labels: L) -> Self {
+        let inner = labels.get_from_metric_vec(vec);
+        Self {
+            inner,
+            labels,
+            vec: vec.clone(),
+            _phantom: &PhantomData,
+        }
+    }
+}
+
+impl<'a, P, L> Deref for DeleteOnDropGauge<'a, P, L>
+where
+    P: Atomic,
+    L: PromLabelsExt<'a>,
+{
+    type Target = GenericGauge<P>;
+    fn deref(&self) -> &GenericGauge<P> {
+        &self.inner
+    }
+}
+
+impl<'a, P, L> Drop for DeleteOnDropGauge<'a, P, L>
+where
+    P: Atomic,
+    L: PromLabelsExt<'a>,
+{
+    fn drop(&mut self) {
+        if self.labels.remove_from_metric_vec(&self.vec).is_err() {
+            // ignore.
+        }
```

Comment:
This implementation removes the ability to introspect that things have gone wrong, I think without fixing https://github.com/MaterializeInc/rust-prometheus/issues/12 which is probably fine because we were ignoring those issues anyway, but we should at least bring that issue over to the materialize repo.

Reply:
Hmmm, that's right - I was not quite sure under what conditions an error would occur! I'll port over the issue, maybe we should only allow creating a single metric instance for any set of labels (which I guess, we now have a record of!)... I'll move the issue over when this goes in (: